### PR TITLE
Group type management

### PR DESCRIPTION
**Pull request by Wisfakim**
It's management of type for groups. 
Each group can be 'unique' so each player can have one group of this type

It's my first contribution, tell me what i'm doing wrong
Tests : 
https://streamable.com/ju27zj

groups_types table 
image

groups table using in test
image

### CODE REVIEW

**Comment by thelindat** [comment]
What’s the actual point? If you don’t want to assign people to multiple “jobs” or “gangs” just don’t do that; more a script level restriction than database. Giving it an arbitrary type is fine but unique 🤷

**Comment by Wisfakim** [comment]
I think it can be offer way to manage the type of groups and prevent player input, like recruit player is already in a gang.
In my mind a lot of roleplay server needs this option.
By the way it's transparent, you can don't set type and it's working like actual.

File: database.sql [hunk 1]
```diff
@@ -65,13 +65,24 @@ CREATE TABLE IF NOT EXISTS `character_inventory` (
   CONSTRAINT `FK_inventory_characters` FOREIGN KEY (`charid`) REFERENCES `characters` (`charid`) ON UPDATE CASCADE ON DELETE CASCADE
 );
 
+CREATE TABLE IF NOT EXISTS `ox_groups_types` (
+  `code` varchar(20) NOT NULL,
```
Comment (thelindat) [inline database.sql:69]: Having the group type as 'code' in the types table is a bit weird.

File: database.sql [hunk 2]
```diff
@@ -65,13 +65,24 @@ CREATE TABLE IF NOT EXISTS `character_inventory` (
...
+  `label` varchar(20) DEFAULT NULL,
```
Comment (thelindat) [inline database.sql:70]: Not sure how useful a type label is necessarily.

File: database.sql [hunk 3]
```diff
@@ -65,13 +65,24 @@ CREATE TABLE IF NOT EXISTS `character_inventory` (
...
+  `unique` int(1) DEFAULT NULL,
```
Comment (thelindat) [inline database.sql:71]: I'd like to use BIT for storing booleans, as it's a bit more clear at a glance.

File: server/groups/class.lua [hunk 4]
```diff
@@ -5,6 +5,9 @@
 ---@field principal string
 ---@field hasAccount boolean
 ---@field adminGrade number
+---@field type = string,
+---@field type_label = string,
```
Comment (thelindat) [inline server/groups/class.lua:9]: type_label either not useful (above) or should be typeLabel (camel casing).
unique_type should be uniqueType, isUnique, or such.

File: server/groups/class.lua [hunk 5]
```diff
@@ -85,6 +88,15 @@ function CGroup:set(player, grade)
     if currentGrade then
         if currentGrade == grade then return end
         self:remove(player, currentGrade)
+    else
+        if self.type and self.unique_type == 1 then
+            for _ in pairs(player.private.groups) do
```
Comment (thelindat) [inline server/groups/class.lua:93]: Weird to use `_` when using the variable. Would also be good to have an exported function or method to check if a group can be added.

File: server/groups/class.lua [hunk 6]
```diff
@@ -85,6 +88,15 @@ function CGroup:set(player, grade)
...
+                local groupCompare = Ox.GetGroup(_)
+                if groupCompare.type == self.type then
+                    error(("Attempted to add multiple unique group '%s' with '%s' for player.%s"):format(self.type, self.name, player.source))
```
Comment (thelindat) [inline server/groups/class.lua:96]: Throwing an error can be a bit awkward when an error didn't actually occur, especially when pcall isn't a common practice in FiveM.

A standardised warning system has finally been added to FiveM so you can use `warn(message)` instead without ending the current thread, then we need to return a boolean for success/failure.

**Review by thelindat — CHANGES_REQUESTED** [review]

**Comment by Wisfakim** [comment]
Changes done, I deleted sql request with LEFT OUTER JOIN, something goes wrong with oxmysql when we got null and join.

**Comment by Wisfakim** [comment]
This edit break nothing of actual groups purpose.
Maybe you are true, is not my mind. Like already said I think lot of roleplaye server need this, and if you got multiple script to setjob, you can make restriction for it at the core level and for me it's pretty good things. 🤷

For a SQL I don't think is syntax error, the join work great on SQL manager, and work great on oxmysql if no one of type are null.

**Comment by thelindat** [comment]
> Why shouldn't a player be able to have multiple of the same "type" of group if you as an admin/owner want them to and if you don't want them to just don't assign them to multiple.

That was my initial thought and is why I haven't merged this yet, however there may be cases where you don't want a player to be in several gangs; perhaps having group types and something like `player.hasGroupType('gang')`?

**Comment by Wisfakim** [comment]
You true Lindat ! It's good solution to implement this feature for external script . I do commit in few days to remove unique column, checkin and write export.

**Pull request closed without merging**